Hw1

arekalov opened the pull request.

arekalov committed:
feat: add ToDoListTheme

Add 3 files:
Color.kt - colors for all color schemes
Type.kt - app
typography
Theme.kt - color scemes, basic app theme.
Also add simple content to MainActivity for test theme.

arekalov committed:
feat: add basic class models

Add entity ToDoItem and repository ToDoItemRepository with fake data.

arekalov committed:
style: add icons

Add icons fo whole app. Remove content for theme demonstration from
MainActivity.

arekalov committed:
feat: add prototype of ToDoItem ui

Create a few composable functions for draw and preview ToDoItems and
toDoItemList

arekalov committed:
feat: add ToDoItemsViewModel

Add states (viewModel, lambdas for using viewModel).
Edit ToDoItem data class for using state.

arekalov committed:
refactor: edit viewModel, add deletind

Change List to Flow in viewModel fix bugs with state.
Add swipe deleting
by both swipes.

arekalov committed:
feat: add detail screen and navigation

EditScreen - detail screen. Also add constansts to navigation. Fix smal
bags.

arekalov committed:
feat: add fab logic

Fab add new item in list.

arekalov committed:
feat: add deadline logic

MAXBAF1 commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItem.kt:
```diff
@@ -0,0 +1,21 @@
+    var isDone: MutableState<Boolean>,
+    var creationDate: MutableState<LocalDate>,
+    var editDate: MutableState<LocalDate?> = mutableStateOf(null),
+) : Serializable
```
Зачем всё делать MutableState? Такого явно не должно быть, у тебя по StateFlow и так будет происходить рекомпозиция. то что ты делаешь слишком излишне

NewUserKK replied:
поддерживаю, можно сделать это всё просто data-моделькой (и ещё сделать Parcelable вместо Serializable)

MAXBAF1 commented on app/src/main/java/com/arekalov/yandexshmr/HomeScreen.kt:
```diff
@@ -0,0 +1,335 @@
+                        creationDate = mutableStateOf(LocalDate.now())
+                    )
+                    toDoItemsViewModel.addItem(item)
+                    onItemClick(item)
```
Ты создаёшь дело при клике на fab и сразу добавляешь его в viewModel, а потом открываешь экран редактирования по этому делу. Из-за этого у тебя не работает кнопка "отменить", в любом случае будет создано как минимум пустое дело. Лучше сделать возможность переходить на экран редактирования/создания без передавания туда дела (можно сделать nullable), и уже внутри создавать новое дело, имея под каждый экран свою viewModel:
1. В случае сохранения, сохранять в репозиторий и переходить на Home экран, там уже брать изменения репоитория.
2. Если пользователь нажимает на крестик, то просто переходим на Home

MAXBAF1 commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -0,0 +1,290 @@
+                mutableStateOf<LocalDate?>(null),
+                mutableStateOf(false),
+                mutableStateOf(LocalDate.of(2024, 7, 15)),
+                mutableStateOf(LocalDate.of(2024, 7, 14))
```
У тебя minSDK = 24, а метод LocalDate.of() требует минимум 26, то есть на 24-ом и 25-ом всё это не будет работать. Проще хранить время в Unix формате, да и с DatePicker'a можно получить Unix время

NewUserKK replied:
можно ещё в старых-добрых Calendar хранить, чтобы было более удобно с ними работать

ещё есть библиотечки, которые бэкпортят функциональность java8-библиотек для дат на старую джаву (и соответственно, на старые андроиды)

MAXBAF1 reviewed (commented):
Привет, посмотрел, основное ТЗ выполнено, но есть недостатки. Понравилось, что если дело становится выполненным, иконка его приоритета становится зелёной.

Замечания по коду ниже, здесь напишу неудобства самого приложения со стороны пользователя:
1.  Чтобы выбрать дедлайн я нажимаю на switch и ожидаю, что передо мной появится эта возможность, но тут устанавливается какая-то дата по умолчанию "2024-06-23" и совсем не понятно как её изменить. Только, попереключав switch несколько раз, я понял что тут он не поможет, оказалось, чтобы поменять дату нужно кликнуть на блок где она находится. При этом, когда у тебя появился DatePickerDialog, если тыкнуть на пустое пространство или нажать "назад", то поменять дату снова уже не получится (DatePickerDialog просто не появляется)
2. Свайп привычки в обе стороны приводит к её удалению, там нет иконки мусорки, поэтому до того как удалишь вообще непонятно что оно удалится. Также фон всего дела при свайпе становится полностью красным, что выглядит некрасиво и приводит к ощущению что, если отпустишь даже оставив на месте, оно всё равно удалится. Ещё закругление красного фона у первого и последнего элемента сильно больше чем у самого контейнера дел.
3. Ну и просто совет как можно сделать topAppBar более красивым (как на макете), у тебя сейчас просто он просто схлапывается. Есть статься на Хабре как сделать красивую анимацию передвижения и уменьшения заголовка в вид topAppBar'a, вот: https://habr.com/ru/companies/hh/articles/703192/

MAXBAF1 commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -0,0 +1,290 @@
+
+    fun addItem(item: ToDoItem) {
+        itemList.add(item)
+    }
```
Эти методы существуют, но не используются

arekalov commented:
> Привет, посмотрел, основное ТЗ выполнено, но есть недостатки. Понравилось, что если дело становится выполненным, иконка его приоритета становится зелёной.
> 
> Замечания по коду ниже, здесь напишу неудобства самого приложения со стороны пользователя:
> 1.  Чтобы выбрать дедлайн я нажимаю на switch и ожидаю, что передо мной появится эта возможность, но тут устанавливается какая-то дата по умолчанию "2024-06-23" и совсем не понятно как её изменить. Только, попереключав switch несколько раз, я понял что тут он не поможет, оказалось, чтобы поменять дату нужно кликнуть на блок где она находится. При этом, когда у тебя появился DatePickerDialog, если тыкнуть на пустое пространство или нажать "назад", то поменять дату снова уже не получится (DatePickerDialog просто не появляется)
> 2. Свайп привычки в обе стороны приводит к её удалению, там нет иконки мусорки, поэтому до того как удалишь вообще непонятно что оно удалится. Также фон всего дела при свайпе становится полностью красным, что выглядит некрасиво и приводит к ощущению что, если отпустишь даже оставив на месте, оно всё равно удалится. Ещё закругление красного фона у первого и последнего элемента сильно больше чем у самого контейнера дел.
> 3. Ну и просто совет как можно сделать topAppBar более красивым (как на макете), у тебя сейчас просто он просто схлапывается. Есть статься на Хабре как сделать красивую анимацию передвижения и уменьшения заголовка в вид topAppBar'a, вот: https://habr.com/ru/companies/hh/articles/703192/

Привет, большое спасибо за такой подробный фитбэк. Хотел спросить про оценку и дать комментарии. 
1. Я посчитал установку даты по умолчанию оптимальным и удобным. Но сейчас я согласен с тобой и исправлю эту логику. 
2. Это вроде как доп задания,я уже немного не успевал и оставил пока так. 
3. За это большое спасибо.

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+    val item = remember {
+        toDoItemsViewModel.getItem(id)
+    }
+    println(id)
```
лучше не оставлять такое дебажное в коде -- а если что-то нужно залогировать, то использовать класс Log

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+    modifier: Modifier = Modifier,
+    toDoItemsViewModel: ToDoItemsViewModel = viewModel(),
+) {
+    val item = remember {
```
вот эта штука получается nullable, а дальше у тебя много где просто unwrap-ы через !!, что выглядит довольно опасно

лучше тогда разделить два варианта, сразу проверить на null и отрисовать соответствующий ui

ещё можно передавать не айдишник, а сразу готовую модельку, чтобы вьюшка просто знала, что ей нужно её отрисовать

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+        topBar = {
+            AppBar(
+                onBack = onBack,
+                onSave = { toDoItemsViewModel.update(item!!.id.value, item); onBack() })
```
nit: не стоит экономить строки написанием через `;`

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+            HorizontalDivider()
+            DeadlinePicker(
+                deadline = item?.deadline?.value,
+                onSwitchChanged = {
```
nit: лучше заименовать всё же

```
item?.deadline?.value = if (checked) LocalDate.now() else null
```

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+import com.arekalov.yandexshmr.ui.ToDoListTheme
+import java.time.LocalDate
+
+@RequiresApi(Build.VERSION_CODES.O)
```
а что если меньше?

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/EditScreen.kt:
```diff
@@ -0,0 +1,370 @@
+    onDateSelected: (LocalDate) -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val mContext = LocalContext.current
```
nit: обычно в котлине префикс `m` не используют

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/HomeScreen.kt:
```diff
@@ -0,0 +1,335 @@
+            onCheckedChange = onCheckedChange,
+            colors = CheckboxDefaults.colors(
+                checkedColor = MaterialTheme.colorScheme.secondary,
+                uncheckedColor = (if (item.priority.value == Priority.HIGH) MaterialTheme.colorScheme.tertiary
```
nit: для лучшей читаемости лучше будет
```
uncheckedColor = if (...) {
  ...
} else {
  ...
}
```

ниже так же

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/HomeScreen.kt:
```diff
@@ -0,0 +1,335 @@
+                else R.drawable.ic_down_arrow
+            ),
+            contentDescription = "priority",
+            tint = if (item.isDone.value && item.priority.value != Priority.REGULAR) MaterialTheme.colorScheme.secondary
```
более читаемо будет заменить на when

NewUserKK commented on app/src/main/java/com/arekalov/yandexshmr/models/ToDoItemRepository.kt:
```diff
@@ -0,0 +1,290 @@
+
+    init {
+        val todoItemsToAdd = listOf(
+            ToDoItem(
```
хорошо бы использовать именованные аргументы, особенно в классах с таким количеством полей

так и меньше вероятность ошибиться в порядке, и ревьюерам проще смотреть (хороший тон -- и в джаве, где нет поддержки именованных аргументов, писать что-то вроде `new DataClass(/*arg1=*/ true, /*arg2=*/ 0)`)

понятно, что здесь это просто копипаста хардкоженных данных, но в целом стоит об этом помнить (что-то похожее, например, может быть при написании тестов, и их проще смотреть и править с явными именованными аргументами)

NewUserKK reviewed (commented):
покомментил немного по коду, на оценку не влияет :)

NewUserKK commented:
@arekalov валидация оценки по запросу

## Баги, которые влияют на функционал по ТЗ
1. Если нажать на крестик в меню добавления/редактирования дела, то экран не закроется, а просто произойдёт новый переход на домашний экран со списком дел вместо того, чтобы удалить экран редактирования из backstack-а навигации. Получается длинный бэкстек, по которому ещё и можно навигироваться по нажатию на системный back.
https://newuserkk.s-ul.eu/NAZiLkZM

2. DatePicker не открывается заново при отмене/клике вовне -- это основная проблема безотносительно его UX и взаимодействий со свичером. https://newuserkk.s-ul.eu/OSaU58N7

4. Креш при запуске на 24-25 API андроида из-за использования LocalDate https://gist.github.com/NewUserKK/5bb7bd7283a838c907c98ed946a39a33 -- причём, студия это даже подсвечивает. Нужно тогда или ставить minApi=26, или гарантировать совместимость.

## Мелкие недочёты
1. Меню выбора приоритета открывается не рядом с кнопкой <img width="250" alt="image" src="https://github.com/arekalov/YandexSHMR/assets/33187027/928de317-efc3-459f-971b-6d170cf73d22">

2. Есть кнопка удаления на несозданное дело
    - в ТЗ конечно явно не прописано, но логично её не показывать, потому что дело ведь ещё не создано

3. Дело создаётся несмотря на "отмену" (должно на нажатие кнопки "сохранить")
    - корень проблемы, как и у предыдущего, тут https://github.com/arekalov/YandexSHMR/pull/1#discussion_r1650026074
    - по ТЗ по факту не требовалось делать логику манипуляции данными, поэтому напрямую на оценку не влияет

4. Касательно UX DatePicker-а -- по факту, можно обойтись вообще без switcher-а рядом, но если уж его делать, то стоит дизейблить нажатие на блок с датой. Но без свичера вообще всё интуитивно будет понятно.

5. Хотелось бы чуть побольше паддинг между иконкой приоритета и текстом дела

## Код
Основная проблема -- использование MutableState<*> как полей в data class. Модель должна быть обычным набором иммутабельных полей, стейт -- это ответственность уже ViewModel и всего остального -- потом эта же модель будет конвертироваться и в модели для слоя сети/базы данных. Из-за этого и в коде экранов получается куча лишнего. Плюсом к этому, все поля data class-а объявлены как var, а не val, что может повлечь баги при рекомпозиции если кто-то будет сетить поле извне напрямую. Обычно у моделек все поля val и модификации происходят через создание нового инстанса через `copy`.

## Доп задания и прочие плюсы
1. Сделана манипуляция данными (добавление/удаление/редактирование), вью на это реагирует через Flow
2. Работает swipe-to-delete, хоть и не суперкрасиво (ненужный ripple, появляется белая стрелочка у обычного приоритета)
3. Полно представлены данные в репозитории
4. Строки вынесены в ресурсы

## Итого
С оценкой согласен, оставляем как есть

arekalov committed:
fix: edit ToDoItem properties to types without MutableState, fix fab bug

MutableState properties are redundant for ToDoItem model, so it's create a set of changes:
edit ToDoItem, edit Repository, edit screens and viewModel.
Fix problem with FAB (now model created in edit screen).
Fix not important bug, clean code (delete println).

arekalov committed:
refactor: structure the code, add new directories

Move navigation to separate file, create new dirs for future scaling.

arekalov committed:
fix: Fix problems with deadline picking

Delete switcher and add icon for only disabling deadline date.
